linkerd/stack/src/optional.rs

```diff
@@ -0,0 +1,40 @@
+use crate::{layer, Either, NewService};
+
+/// A stack middleware that takes an optional target type and builds an `A`-typed
+/// stack if the target is set and a `B`-typed stack if it is not.
+#[derive(Clone, Debug, Default)]
+pub struct NewOptional<A, B> {
+    new_some: A,
+    new_none: B,
+}
```

Comment:
Isn't this just `MakeSwitch` with a predicate like this?
```rust
MakeSwitch::new(new_some, new_none, |&(ref t, _)| t.is_some())
```

It's not a big deal to me, but it seems like we could just use the existing `MakeSwitch` layer for this (it already has a `NewService` impl). If you think it's worth having a separate `NewOptional` type, it could just be a wrapper around `MakeSwitch`. 

The code is simple enough that it's not a *huge* deal either way, just a thought.

Reply:
no because it unwraps the option -- the inner services are NewService<(T, U)> and NewService<U> instead of both being NewService<(Option<T>, U)>, which is what would be needed for MakeSwitch

Reply:
oh, yup, i missed that --- this does make sense as a separate thing, then